crypto/bn/bn_mpi.c

```diff
@@ -45,7 +45,7 @@ BIGNUM *BN_mpi2bn(const unsigned char *d, int n, BIGNUM *ain)
     int neg = 0;
     BIGNUM *a = NULL;
 
-    if (n < 4) {
+    if (n < 4 || d[0] & 0x80) {
```

Comment:
this looks wrong

Reply:
Why? Could you be please more specific? I am trying to avoid values which could not be represented in signed long type (after the shift) which is used in the following code.

Reply:
First long is usually longer than 4 bytes.
But also d[0..3] is probably not a signed quantity,
d[4] is the sign bit.

Reply:
I think the problem is it uses long, probably unsigned long would be more appropriate?

Reply:
could it be that this is not the problem you wanted to address, since this code
has lots more undefined behavior than this?

Reply:
Wouldn't we want to compare the result of the bitwise-and against zero anyway, as a matter of style?

Reply:
to be completely correct you should add `&& sizeof(long) == 4`

Reply:
So we want to support a more than 2GB large BIGNUMs on systems that allow this? I think it is better to be consistent and have the limit be the same on all supported architectures.

Reply:
good argument but dor a denial of service attack also 100.000 bit BIGNUMS would
be a threat.

Reply:
> good argument but dor a denial of service attack also 100.000 bit BIGNUMS would
> be a threat.

Now that would be for a different PR. But I do not think this function is used in a way that a DoS attack would apply.

Reply:
you are of course right that this needs improvements.
But if you change the de-serialization to choke at 2GB you should
also change the inverse function above.  IT has undefined behavior
at 2GB, due to integer overflow.

Reply:
any way, it is fun for me to discuss these issues with you.
and disagreement is not per se a bad thing.